### DISCUSSION TITLE
Is it possible to extend a YesSql query after calling .ListAsync()?

### DISCUSSION THREAD

**Discussion by netwavebe** [OrchardCMS/OrchardCore#13527] [Q&A]
Assume I have a query like this:
`var query = _session.Query<ContentItem, ContentItemIndex>().Where(c => c.ContentType == "Country");`

I want to get the results of this query:
`var countries1 = await query.ListAsync();`

Next I need to extend the same query:
`query = query.With<ContentItemIndex>().Where(c => c.Published);`

But when I try to get the results of the extended query, the results are wrong:
`var countries2 = await query.ListAsync();`

When I remove the line to get `countries1` (so remove ListAsync()) , `countries2` is correct.

So after calling ListAsync on a query, everything that's done to extend the query is ignored. Is there a way to work around this problem? Perhaps clone a query before calling ListAsync()? What are my options?

**Comment by hishamco** [comment]
I don't think so, because of 

https://github.com/sebastienros/yessql/blob/c0771eb555c67bf419859016cb3563a3540fe7c7/src/YesSql.Core/Services/DefaultQuery.cs#L1239-L1240

https://github.com/sebastienros/yessql/blob/c0771eb555c67bf419859016cb3563a3540fe7c7/src/YesSql.Core/Services/DefaultQuery.cs#L1247

@sebastienros correct me if I'm wrong

**Reply by netwavebe** [reply]
@hishamco: Thanks for taking the time to comment!

Is there a way to clone the query before calling ListAsync()? Perhaps @sebastienros knows a way to work around this issue?

**Reply by hishamco** [reply]
If there's a way it might works as client query on the memory something similar to `IEnumerable`. 

Waiting for Seb's comment

**Comment by ns8482e** [comment]
ListAsync executes query and returns data, everything after ListAsync is just c# objects and you can use Linq to filter

**Reply by netwavebe** [reply]
I think you misunderstood my question. I want to extend the query ListAsync was called on.

**Reply by ns8482e** [reply]
I understand! I meant regardless of extending second query - you are already executing first query so there is no much benefit extending second query and run on SQL - rather you already have result from first query - just filter using linq or plain C# foreach

**Comment by sebastienros** [comment]
I assume one could expect it to work, so probably should be treated as a bug. It's possible that calling ListAync is mutating the query internally and could instead clone the state before doing so, or do it in a way that wouldn't change the underlying constraints. Something to test in yessql's repository directly and fix there.
